### PR TITLE
[oree-834] Resolving sdk.authorize() promise

### PR DESCRIPTION
**Pull request by nimal-outreach**
SDK part of the fix for a bug of authorization promise not being resolved after the OAuth compeltes

### CODE REVIEW

**Review by honeydripper — CHANGES_REQUESTED** [review]
Please add tests

**Comment by nimal-outreach** [comment]
@honeydripper  Tests are added - please review them

File: tests/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
+import { ExtensibilitySdk, Message, OAuthDialogCompletedMessage } from '../src/index';
+import runtime from '../src/sdk/RuntimeContext';
+import oauthService from '../src/sdk/services/oauthService';
+
+describe('sdk tests', () => {
+  let sdk: ExtensibilitySdk;
+
+  describe.only('receiving messages', () => {
```
Comment (honeydripper) [inline tests/index.test.ts:8]: Is the `.only` bit correct?

File: tests/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+      originalOpenPopup = oauthService.openPopup;
+      oauthService.openPopup = jest.fn();
+
+      originalEventListener = window.addEventListener;
+      window.addEventListener = (event: string, handler: EventListenerOrEventListenerObject): void => {
+        if (event === 'message') {
+          messageHandler = handler;
+        }
+      };
+      runtime.application = {
```
Comment (honeydripper) [inline tests/index.test.ts:25]: `.application` and `.runtime` is overridden here but not restored in `afterEach` is that correct?

File: tests/index.test.ts (hunk 1 above)
Comment (palootcenas-outreach) [inline tests/index.test.ts:8]: It should be probably w/o `only`
```suggestion
  describe('receiving messages', () => {
```

File: tests/index.test.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    let originalOpenPopup: (redirectUri?: string, state?: { [key: string]: string }) => void;
+
+    beforeEach(async () => {
+      window.setTimeout = jest.fn().mockImplementation(() => 123) as any;
+
+      originalOpenPopup = oauthService.openPopup;
+      oauthService.openPopup = jest.fn();
+
+      originalEventListener = window.addEventListener;
+      window.addEventListener = (event: string, handler: EventListenerOrEventListenerObject): void => {
```
Comment (honeydripper) [inline tests/index.test.ts:20]: you are overwriting the `addEventListener` for all events but not processing events other than `message`

File: tests/index.test.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+describe('sdk tests', () => {
+  let sdk: ExtensibilitySdk;
+
+  describe.only('receiving messages', () => {
+    var messageHandler: EventListenerOrEventListenerObject;
+    let originalEventListener: (event: string, handler: EventListenerOrEventListenerObject) => void;
+    let originalOpenPopup: (redirectUri?: string, state?: { [key: string]: string }) => void;
+
+    beforeEach(async () => {
+      window.setTimeout = jest.fn().mockImplementation(() => 123) as any;
```
Comment (palootcenas-outreach) [inline tests/index.test.ts:14]: Maybe worth resetting it to original `setTimeout` in `afterEach` ?

File: tests/index.test.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+      runtime.application = {
+        api: {
+          client: { id: '' },
+          redirectUris: [],
+          scopes: [],
+        },
+      } as any;
+      runtime.origin = 'test.outreach.io';
+      sdk = new ExtensibilitySdk();
+      sdk.init();
```
Comment (honeydripper) [inline tests/index.test.ts:34]: you call this before every test. Does it have sideeffects or are we certain it's pure and will remain so?

File: tests/index.test.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    });
+
+    afterEach(() => {
+      window.addEventListener = originalEventListener;
+      oauthService.openPopup = originalOpenPopup;
+      jest.restoreAllMocks();
+    });
+
+    it('will register message event listener', () => {
```
Comment (honeydripper) [inline tests/index.test.ts:43]: Could we make the name a bit more descriptive?

File: tests/index.test.ts [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+      it('will reject auth promise when success', async () => {
+        var message = new OAuthDialogCompletedMessage();
+        message.result = '401';
+        handleMessage(message);
+
+        await expect(authPromise).rejects.toEqual({ message: undefined, result: '401' });
+      });
+    });
+
+    const handleMessage = (message: Message) => {
```
Comment (honeydripper) [inline tests/index.test.ts:70]: nit: suggest moving this to the top

File: tests/index.test.ts [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+      if (handlerEventListenerObject(messageHandler)) {
+        messageHandler.handleEvent(eventMessage);
+      } else {
+        messageHandler(eventMessage);
+      }
+    };
+  });
+});
+
+function handlerEventListenerObject(
```
Comment (honeydripper) [inline tests/index.test.ts:85]: nit: I'd drop this function it does not seem to do much.

File: tests/index.test.ts (hunk 3 above)
Comment (honeydripper) [inline tests/index.test.ts:20]: `window.addEventListener` typically accepts a function as second parameter, why is the signature different here?

**Review by honeydripper — CHANGES_REQUESTED** [review]

File: tests/index.test.ts (hunk 1 above)
Comment (nimal-outreach) [reply tests/index.test.ts]: removed

File: tests/index.test.ts (hunk 6 above)
Comment (nimal-outreach) [reply tests/index.test.ts:52]: restored

File: tests/index.test.ts (hunk 2 above)
Comment (nimal-outreach) [reply tests/index.test.ts:30]: restored

File: tests/index.test.ts (hunk 3 above)
Comment (nimal-outreach) [reply tests/index.test.ts:23]: I don't know, it is definition of the event handler in the typescript which I guess is based on html/dom standards

File: tests/index.test.ts (hunk 5 above)
Comment (nimal-outreach) [reply tests/index.test.ts:39]: In order for our SDK to do anything,.init has to be called - every method checks for that before being executed so init has to be ran. Due to the previous line where a new instance of SDK is created for every test there is no side effects

File: tests/index.test.ts (hunk 3 above)
Comment (nimal-outreach) [reply tests/index.test.ts:23]: Our SDK is using only message event listener - added throwing of an error if event name is not a "message"

File: tests/index.test.ts (hunk 1 above)
Comment (nimal-outreach) [reply tests/index.test.ts]: removed

File: tests/index.test.ts (hunk 4 above)
Comment (nimal-outreach) [reply tests/index.test.ts:17]: restored

File: tests/index.test.ts (hunk 7 above)
Comment (nimal-outreach) [reply tests/index.test.ts:81]: I like keeping helper functions on the bottom as they are secondary in importance to the tests itself

File: tests/index.test.ts (hunk 8 above)
Comment (nimal-outreach) [reply tests/index.test.ts:96]: dropped

File: tests/index.test.ts (hunk 6 above)
Comment (honeydripper) [reply tests/index.test.ts:54]: setting event listener is implementation detail. For which purpose is it set? The name should reflect that.

File: tests/index.test.ts (hunk 3 above)
Comment (honeydripper) [reply tests/index.test.ts:23]: oh wow, I did not know that submitting objects is acceptable in the standard. https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener Sorry 👍

File: tests/index.test.ts (hunk 6 above)
Comment (nimal-outreach) [reply tests/index.test.ts:54]: purpose is to verify that event listener is registered

You know what - I will just remove this test

Comment (nimal-outreach) [reply tests/index.test.ts]: removed

**Comment by nimal-outreach** [comment]
@palootcenas-outreach is there anything else you suggest changing in order to approve the PR?

**Review by palootcenas-outreach — APPROVED** [review]

**Pull request merged**